plugin-server/src/utils/rate-limiter.ts

```diff
@@ -0,0 +1,37 @@
+import { DateTime } from 'luxon'
+
+import { DB } from './db/db'
+
+export type RateLimitedResource = 'events' | 'recordings'
+
+const CACHE_TTL_SECONDS = 60
+export const RATE_LIMITER_CACHE_KEY = '@posthog-plugin-server/rate-limiter/'
+
+export class RateLimiter {
+    private db: DB
+    public localCache: { [key: string]: string[] } = {}
+    public localCacheExpiresAt: DateTime | null = null
+
+    constructor(db: DB) {
+        this.db = db
+    }
+
+    private async refreshCaches(): Promise<void> {
+        if (!this.localCacheExpiresAt || DateTime.utc() > this.localCacheExpiresAt) {
+            const [events, recordings] = await Promise.all([
+                this.db.redisZRange(`${RATE_LIMITER_CACHE_KEY}events`, 0, -1),
+                this.db.redisZRange(`${RATE_LIMITER_CACHE_KEY}recordings`, 0, -1),
+            ])
+            this.localCache = {
+                events: events || [],
+                recordings: recordings || [],
+            }
+            this.localCacheExpiresAt = DateTime.utc().plus({ seconds: CACHE_TTL_SECONDS })
+        }
+    }
```

Comment:
I notice you got rid of the local promise object. If 20 requests came in at once, would this not trigger 40 parallel redisZrange calls that are all doing the same thing?